File: PortTunnel Android/CSharp/MainActivity.cs
/*
MIT License
Copyright (c) 2019 Patrick72762
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

using Android.App;
using Android.Content;
using Android.Content.PM;
using Android.OS;
using Android.Support.V7.App;
using Android.Widget;
using PortTunnelLib;

namespace PortTunnel_Android
{
    [Activity(Label = "@string/app_name", Theme = "@style/AppTheme", MainLauncher = true, ScreenOrientation = ScreenOrientation.Portrait)]
    public class MainActivity : AppCompatActivity
    {
        EditText etH1, etH2, etP1, etP2, etCP, etTo;
        TextView log, activeConnections;
        Button btnStart, btnStop;
        CheckBox cbHttp, cbSsl;
        private PortTunnel tunnel;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);

            // Set our view from the "main" layout resource
            SetContentView(Resource.Layout.activity_main);
            
            GetItems();
            SetActions();
        }

        void GetItems()
        {
            etH1 = FindViewById<EditText>(Resource.Id.etHost1);
            etH2 = FindViewById<EditText>(Resource.Id.etHost2);
            etP1 = FindViewById<EditText>(Resource.Id.etPort1);
            etP2 = FindViewById<EditText>(Resource.Id.etPort2);
            etCP = FindViewById<EditText>(Resource.Id.etControlPort);
            etTo = FindViewById<EditText>(Resource.Id.etTimeout);

            log = FindViewById<TextView>(Resource.Id.txtLog);
            activeConnections = FindViewById<TextView>(Resource.Id.txtActiveConnections);

            btnStart = FindViewById<Button>(Resource.Id.btnStart);
            btnStop = FindViewById<Button>(Resource.Id.btnStop);

            cbHttp = FindViewById<CheckBox>(Resource.Id.cbHttp);
            cbSsl = FindViewById<CheckBox>(Resource.Id.cbSsl);

            ISharedPreferences preferences = GetSharedPreferences("SETUP", FileCreationMode.Private);

            etH1.Text = preferences.GetString("H1", "illumina-chemie.de");
            etH2.Text = preferences.GetString("H2", "");

            etP1.Text = preferences.GetInt("P1", 80).ToString();
            etP2.Text = preferences.GetInt("P2", 8080).ToString();

            etCP.Text = preferences.GetInt("CP", 0).ToString();
            etTo.Text = preferences.GetInt("TO", 1000).ToString();

            cbHttp.Checked = preferences.GetBoolean("HTTP", false);
            cbSsl.Checked = preferences.GetBoolean("SSL", false);
        }

        void SetActions()
        {
            btnStart.Click += BtnStart_Click;
            btnStop.Click += BtnStop_Click;
        }

        private void BtnStop_Click(object sender, System.EventArgs e)
        {
            try
            {
                tunnel.Stop();
            }
            catch { }

            try
            {
                PortTunnel.ExitEnvironment();
            }
            catch { }
        }

        private void BtnStart_Click(object sender, System.EventArgs e)
        {
            try
            {
                string h1 = etH1.Text;
                string h2 = etH2.Text;

                int p1 = int.Parse(etP1.Text);
                int p2 = int.Parse(etP2.Text);

                int cp = int.Parse(etCP.Text);
                int to = int.Parse(etTo.Text);

                ISharedPreferences preferences = GetSharedPreferences("SETUP", FileCreationMode.Private);
                ISharedPreferencesEditor editor = preferences.Edit();

                editor.PutString("H1", h1);
                editor.PutString("H2", h2);

                editor.PutInt("P1", p1);
                editor.PutInt("P2", p2);
                editor.PutInt("CP", cp);
                editor.PutInt("TO", to);

                editor.PutBoolean("HTTP", cbHttp.Checked);
                editor.PutBoolean("SSL", cbSsl.Checked);

                editor.Commit();

                Connection c1 = new Connection(h1, p1);
                Connection c2 = new Connection(h2, p2);

                tunnel = new PortTunnel(c1, c2, to, cp)
                {
                    HttpProtocol = cbHttp.Checked,
                    DestIsSsl = cbSsl.Checked
                };

                tunnel.NewLogOutput += (sen, text) =>
                {
                    try
                    {
                        log.Text += text;

                        activeConnections.Text = ((PortTunnel)tunnel).ActiveConnections.ToString();
                    }
                    catch { }
                };

                tunnel.NewStreamLogCompleted += (sen, content) => {

                };

                BlockInputs();

                tunnel.Start();
            }
            catch { }
        }

        private void BlockInputs()
        {
            btnStart.Enabled = false;
            etH1.Enabled=false;
            etH2.Enabled=false;
            etP1.Enabled=false;
            etP2.Enabled=false;
            etCP.Enabled=false;
            etTo.Enabled=false;
            cbHttp.Enabled=false;
            cbSsl.Enabled=false;
        }
    }
}

File: PortTunnel Android/Java/MainActivity.java
package de.patricklutz.porttunnelandroid;

import android.content.SharedPreferences;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;

import PortTunnelLib.Connection;
import PortTunnelLib.LogListener;
import PortTunnelLib.PortTunnel;

public class MainActivity extends AppCompatActivity {
    EditText etH1, etH2, etP1, etP2, etCP, etTo;
    TextView log, activeConnections;
    Button btnStart, btnStop;
    CheckBox cbHttp, cbSsl;
    private PortTunnel tunnel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        getItems();
        setActions();
    }

    void getItems()
    {
        etH1 = findViewById(R.id.etHost1);
        etH2 = findViewById(R.id.etHost2);
        etP1 = findViewById(R.id.etPort1);
        etP2 = findViewById(R.id.etPort2);
        etCP = findViewById(R.id.etControlPort);
        etTo = findViewById(R.id.etTimeout);

        log = findViewById(R.id.txtLog);
        activeConnections = findViewById(R.id.txtActiveConnections);

        btnStart = findViewById(R.id.btnStart);
        btnStop = findViewById(R.id.btnStop);

        cbHttp = findViewById(R.id.cbHttp);
        cbSsl = findViewById(R.id.cbSsl);

        SharedPreferences preferences = getSharedPreferences("SETUP", MODE_PRIVATE);

        etH1.setText(preferences.getString("H1", "illumina-chemie.de"));
        etH2.setText(preferences.getString("H2", ""));

        etP1.setText(Integer.toString(preferences.getInt("P1", 80)));
        etP2.setText(Integer.toString(preferences.getInt("P2", 8080)));

        etCP.setText(Integer.toString(preferences.getInt("CP", 0)));
        etTo.setText(Integer.toString(preferences.getInt("TO", 1000)));

        cbHttp.setChecked(preferences.getBoolean("HTTP", false));
        cbSsl.setChecked(preferences.getBoolean("SSL", false));
    }

    void setActions(){
        btnStart.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startPortTunnel();
            }
        });

        btnStop.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                stopPortTunnel();
            }
        });
    }

    private void stopPortTunnel(){
        try
        {
            tunnel.Stop();
        }
        catch (Exception e)
        { }

        try
        {
            PortTunnel.ExitEnvironment();
        }
        catch (Exception e)
        { }
    }

    private void startPortTunnel(){
        try
        {
            String h1 = etH1.getText().toString();
            String h2 = etH2.getText().toString();

            int p1 = Integer.parseInt(etP1.getText().toString());
            int p2 = Integer.parseInt(etP2.getText().toString());

            int cp = Integer.parseInt(etCP.getText().toString());
            int to = Integer.parseInt(etTo.getText().toString());

            SharedPreferences preferences = getSharedPreferences("SETUP", MODE_PRIVATE);
            SharedPreferences.Editor editor = preferences.edit();

            editor.putString("H1", h1);
            editor.putString("H2", h2);

            editor.putInt("P1", p1);
            editor.putInt("P2", p2);
            editor.putInt("CP", cp);
            editor.putInt("TO", to);

            editor.putBoolean("HTTP", cbHttp.isChecked());
            editor.putBoolean("SSL", cbSsl.isChecked());

            editor.commit();

            Connection c1 = new Connection(h1, p1);
            Connection c2 = new Connection(h2, p2);

            tunnel = new PortTunnel(c1, c2, to, cp);

            tunnel.setHttpProtocol(cbHttp.isChecked());
            tunnel.setDestIsSsl(cbSsl.isChecked());

            tunnel.AddListener(new LogListener() {
                @Override
                public void NewLogOutput(String s) {
                    final String text = s;
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            try
                            {
                                log.setText(log.getText().toString() + text);

                                activeConnections.setText(((PortTunnel)tunnel).getActiveConnections());
                            }
                            catch (Exception ex)
                            { }
                        }
                    });
                }
            });

            blockInputs();

            tunnel.Start();
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    private void blockInputs(){
        btnStart.setEnabled(false);
        etH1.setEnabled(false);
        etH2.setEnabled(false);
        etP1.setEnabled(false);
        etP2.setEnabled(false);
        etCP.setEnabled(false);
        etTo.setEnabled(false);
        cbHttp.setEnabled(false);
        cbSsl.setEnabled(false);
    }
}
